### ISSUE TITLE
[YCQL] Run operations within a transaction block in serial

### ISSUE THREAD

**Issue by ameyb** [yugabyte/yugabyte-db#5026]
This is a test with simple YCQL table, performing DELETE and INSERT on the same key in a transaction block  :

```
CREATE TABLE ybdemo_keyspace.test_wl (
    h int,
    r int,
    v int,
    PRIMARY KEY (h, r)
) WITH CLUSTERING ORDER BY (r ASC)
    AND default_time_to_live = 0
    AND transactions = {'enabled': 'true'};
```

**Scenario 1: Works as expected**

```
cqlsh:ybdemo_keyspace> begin transaction 
delete from test_wl where h =1; 
insert into test_wl (h,r,v) values (1,1,1); 
end transaction;
```

**Scenario 2**: Same TXn with clause - **if not exists else error** throws an error...

```
cqlsh:ybdemo_keyspace> begin transaction 
delete from test_wl where h =1; 
insert into test_wl (h,r,v) values (1,1,1) if not exists else error; 
end transaction;
#############################
[12:17 PM] InvalidRequest: Error from server: code=2200 [Invalid query] message="Execution Error. Condition on table test_wl was not satisfied.
begin transaction delete from test_wl where h =1; insert into test_wl (h,r,v) values (1,1,1) if not exists else error; end transaction;
                                                         ^^^^
 (ql error -300)"
```
`Scenario 3:` Variation of Scenario #1 by only providing values for a & b; and not column c; This confirms that value of column c is set to NULL and old value of c does not resurface

```
Initial State
ycqlsh:ybdemo_keyspace> select * from test_wl;
 h | r | v
---+---+-----
 1 | 3 | 110
 2 | 1 |   2
(2 rows)
Final State
ycqlsh:ybdemo_keyspace> begin transaction
                   ...    delete from test_wl where h =1;
                   ...   insert into test_wl(h,r) values (1,3);
                   ... end transaction;
ycqlsh:ybdemo_keyspace> select * from test_wl;
 h | r | v
---+---+------
 1 | 3 | null
 2 | 1 |    2
```

**Comment by kmuthukk** [comment]
Variant test case (where collision is on index) - thx @m-iancu .

```
create keyspace test;
use test;
create table test(h int, r int, v int, primary key (h, r)) with transactions = {'enabled' : true};
create unique index on test(v);
insert into test(h,r ,v) values (1, 1, 99);
begin transaction
delete from test where h = 1 and r = 2;
insert into test(h,r,v) values (1, 3,99);
end transaction;
```

The txn fails with:
```
InvalidRequest: Error from server: code=2200 [Invalid query] message="Execution Error. Duplicate value disallowed by unique index test_v_idx
begin transaction
delete from test where h = 1 and r =2;
insert into test(h,r,v) values (1, 3,99);
       ^^^^
end transaction;
 (ql error -300)"
```

CC: @mbautin

**Comment by m-iancu** [comment]
A repro with indexes (probably the same root cause). 
```
create keyspace test;
use test;
create table test(h int, r int, v int, primary key (h, r)) with transactions = {'enabled' : true};
create unique index on test(v);

insert into test(h,r ,v) values (1, 1, 99);

begin transaction
delete from test where h = 1 and r = 1;
insert into test(h,r,v) values (1, 2, 99);
end transaction;
```
The txn above tries to do a `delete 99` on the index and an `insert 99` right after in the same txn.
It (wrongly) fails with this error:
```
InvalidRequest: Error from server: code=2200 [Invalid query] message="Execution Error. Duplicate value disallowed by unique index test_v_idx
begin transaction
delete from test where h = 1 and r = 1;
insert into test(h,r,v) values (1, 2, 99);
       ^^^^
end transaction;
 (ql error -300)"
```

**Comment by tedyu** [comment]
In Tablet::CompleteQLWriteBatch(), maybe we can iterate QLWriteOperation's and see if there is another operation of QLWriteRequestPB::QL_STMT_INSERT with same value for index ColumnId(s) as the underlying QLWriteRequestPB::QL_STMT_INSERT operation. In the above example, we're interested in value 99 for column v.

Here is some draft code for retrieving index ColumnId(s) (stored in column_refs).
```
      std::vector<ColumnId> index_ids = metadata_.index_key_column_ids();
      yb::QLReferencedColumnsPB& column_pbs = ql_write_op->request().column_refs();
      vector<ColumnId> column_refs;
      for (int32_t id : column_pbs.static_ids()) {
        if (find(index_ids.begin(), index_ids.end(), id) != index_ids.end()) {
          column_refs.emplace_back(id);
        }
      }
      for (int32_t id : column_pbs.ids()) {
        if (find(index_ids.begin(), index_ids.end(), id) != index_ids.end()) {
          column_refs.emplace_back(id);
        }
      }
```
If there is no such (other) insert operation, we can transform the underlying insert operation to update (if it follows delete).

I am in early stage of understanding related code. Please let me know if there is something I missed.

Thanks

**Comment by tedyu** [comment]
Should have looked at QLWriteOperation::HasDuplicateUniqueIndexValue() where duplicate check is done against the associated indexed primary key.
In the example, (h, r) == (1, 2) is different from (1, 1) in the table, hence triggering duplicate check.

Trying to see how the delete row operation can be retrieved from DocOperationApplyData.

**Comment by tedyu** [comment]
Another case:
```
begin transaction
delete from test where h = 1;
insert into test(h,r,v) values (1, 2, 99);
```
The deletion may affect many rows, including the row with (h, r) == (1, 1).
In this case we shouldn't remember the deleted rows in QLWriteOperation (when we see the delete operation) because of the potentially large memory consumption.

**Comment by tedyu** [comment]
From [#3718] Ignore writes from the same transaction while updating CQL indexes (brought up in the discussion for this issue):
```
it would not re-issue the deletion of the 2 => 1 entry from the index.
```
Maybe the modification to index table (deletion in this case) can be cached so that in case of retry, index table modification can be retried as well.

**Comment by m-iancu** [comment]
After internal discussion, looks like the issue might be related to the internal batching/parallelization we do for batches and transaction blocks. The current idea is to just run operations serially for transactions and do not attempt to parallelize (see details below).

Parallelism (i.e. decision when to flush) is done in [executor.cc:AddOperation](https://github.com/yugabyte/yugabyte-db/blob/master/src/yb/yql/cql/ql/exec/executor.cc#L2153) where we add operations in order, but only apply then to the session if there is no dependency (based on the checking done by the write_batch utility). If there is a dependency, the op is effectively postponed and I think it will be handled later in `ProcessAsyncResults` after `FlushAsyncDone`.
So the initial `FlushAsync` at the end of the user statement (txn in this case) should only flush/write the applied (i.e. non-postponed) statements.

The dependency-checking code starts from [executor.cc:WriteBatch::Add](https://github.com/yugabyte/yugabyte-db/blob/master/src/yb/yql/cql/ql/exec/executor.cc#L2113).

If we don't do this checking correctly (which I think we do not for indexes) that means we will try to run some statements in parallel that we should not.

This would explain the index example above, but I don't think it explains the table-only example. Specifically for:
```
begin transaction  
delete from test_wl where h = 1 and r = 1 if exists else error;  
insert into test_wl (h,r,v) values (1,1,1) if not exists else error;  
end transaction;
```
the statements do actually depend on each other (w.r.t to the primary key checking).

So there is likely some other complication here that needs further investigation.

I think we should first confirm that the `AddOperation` code above does get called for transactions and check what it returns.
Then move forward depending on the result of that investigation.

Also, the tests for this feature are in [TestTransaction.java](https://github.com/yugabyte/yugabyte-db/blob/master/java/yb-cql/src/test/java/org/yb/cql/TestTransaction.java) -- we should add the above examples as test cases there once the feature is done.

cc @tedyu @mbautin @kmuthukk

**Comment by tedyu** [comment]
I have added the a test - testInsertAfterDelete.
Here is snippet from test output:
```
[ERROR] testInsertAfterDelete(org.yb.cql.TestTransaction)  Time elapsed: 6.143 s  <<< ERROR!
com.datastax.driver.core.exceptions.InvalidQueryException:
Execution Error. Condition on table test_wl was not satisfied.
begin transaction delete from test_wl where h = 1 and r = 1 if exists else error; insert into test_wl (h,r,v) values (1,1,2) if not exists else error;end transaction;
                                                                                         ^^^^
 (ql error -300)
	at org.yb.cql.TestTransaction.testInsertAfterDelete(TestTransaction.java:294)
Caused by: com.datastax.driver.core.exceptions.InvalidQueryException:
Execution Error. Condition on table test_wl was not satisfied.
begin transaction delete from test_wl where h = 1 and r = 1 if exists else error; insert into test_wl (h,r,v) values (1,1,2) if not exists else error;end transaction;
                                                                                         ^^^^
```
Here is snippet of java/yb-cql/target/surefire-reports_org.yb.cql.TestTransaction__testInsertAfterDelete/org.yb.cql.TestTransaction-output.txt
```
ts2|pid93158|:25385|http://127.0.0.2:12861 I0814 16:52:47.986851 171515904 executor.cc:1784] flushing
ts2|pid93158|:25385|http://127.0.0.2:12861 I0814 16:52:47.987287 169906176 executor.cc:1784] flushing
ts2|pid93158|:25385|http://127.0.0.2:12861 I0814 16:52:47.990963 173125632 executor.cc:1784] flushing
ts2|pid93158|:25385|http://127.0.0.2:12861 I0814 16:52:47.996165 173662208 thread_pool.cc:171] Starting thread pool { name: TransactionManager queue_limit: 150 max_workers: 50 }
ts2|pid93158|:25385|http://127.0.0.2:12861 I0814 16:52:47.996660 173662208 executor.cc:2118] type: 3 columns: 0
ts2|pid93158|:25385|http://127.0.0.2:12861 I0814 16:52:47.996672 173662208 executor.cc:2122]  reads primary 0, primary count 0, hash 0
ts2|pid93158|:25385|http://127.0.0.2:12861 I0814 16:52:47.996695 173662208 executor.cc:2124] 0x11c4193b0 reads static 0 hash key count 0
ts2|pid93158|:25385|http://127.0.0.2:12861 I0814 16:52:47.996704 173662208 executor.cc:2170] the op 0x11c4193b0 added: 1
ts2|pid93158|:25385|http://127.0.0.2:12861 I0814 16:52:47.996781 173662208 executor.cc:2118] type: 1 columns: 1
ts2|pid93158|:25385|http://127.0.0.2:12861 I0814 16:52:47.996789 173662208 executor.cc:2122]  reads primary 0, primary count 0, hash 0
ts2|pid93158|:25385|http://127.0.0.2:12861 I0814 16:52:47.996795 173662208 executor.cc:2124] 0x11c419ae0 reads static 0 hash key count 0
ts2|pid93158|:25385|http://127.0.0.2:12861 I0814 16:52:47.996806 173662208 executor.cc:2170] the op 0x11c419ae0 added: 1
ts2|pid93158|:25385|http://127.0.0.2:12861 I0814 16:52:48.003264 170979328 executor.cc:1784] flushing
ts2|pid93158|:25385|http://127.0.0.2:12861 I0814 16:52:48.004024 170442752 executor.cc:1784] flushing
ts1|pid93157|:18015|http://127.0.0.5:27349 I0814 16:52:48.006170 217280512 thread_pool.cc:171] Starting thread pool { name: TransactionManager queue_limit: 150 max_workers: 50 }
ts1|pid93157|:18015|http://127.0.0.5:27349 I0814 16:52:48.006685 217280512 executor.cc:2118] type: 3 columns: 0
ts1|pid93157|:18015|http://127.0.0.5:27349 I0814 16:52:48.006700 217280512 executor.cc:2122]  reads primary 1, primary count 0, hash 0
ts1|pid93157|:18015|http://127.0.0.5:27349 I0814 16:52:48.006722 217280512 executor.cc:2124] 0x11e2395e0 reads static 0 hash key count 0
ts1|pid93157|:18015|http://127.0.0.5:27349 I0814 16:52:48.006736 217280512 executor.cc:2170] the op 0x11e2395e0 added: 1
ts1|pid93157|:18015|http://127.0.0.5:27349 I0814 16:52:48.006821 217280512 executor.cc:2118] type: 1 columns: 1
ts1|pid93157|:18015|http://127.0.0.5:27349 I0814 16:52:48.006830 217280512 executor.cc:2122]  reads primary 1, primary count 0, hash 0
ts1|pid93157|:18015|http://127.0.0.5:27349 I0814 16:52:48.006836 217280512 executor.cc:2124] 0x11c885cc0 reads static 0 hash key count 0
ts1|pid93157|:18015|http://127.0.0.5:27349 I0814 16:52:48.006850 217280512 executor.cc:2170] the op 0x11c885cc0 added: 1
```
where type 1 is insert, type 3 is delete.
From the snippet, we can see that even when op->ReadsPrimaryRow() is true, ops_by_primary_key_.count(op) is 0 for both delete and insert statements, resulting in Executor::WriteBatch::Add() returning true.
The op (delete and insert) is applied to the session for all 4 statements.

**Comment by tedyu** [comment]
Here is the patch adding the test:
[test-cql-trans.txt](https://github.com/yugabyte/yugabyte-db/files/5077866/test-cql-trans.txt)

Here is the patch adding the log for Executor:
[add-op-log.txt](https://github.com/yugabyte/yugabyte-db/files/5077867/add-op-log.txt)

**Comment by tedyu** [comment]
Tried another way of obtaining hashcode:
```
  auto hc = std::hash<YBqlWriteOp>()(*op);
```
but got:
```
src/yb/yql/cql/ql/exec/executor.cc:2126:13: error: call to implicitly-deleted default constructor of 'std::hash<YBqlWriteOp>'
  auto hc = std::hash<YBqlWriteOp>()(*op);
            ^
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/utility:1573:36: note: default constructor of 'hash<yb::client::YBqlWriteOp>' is implicitly deleted because base class '__enum_hash<yb::client::YBqlWriteOp>' has a deleted default constructor
struct _LIBCPP_TEMPLATE_VIS hash : public __enum_hash<_Tp>
                                   ^
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/utility:1567:5: note: '__enum_hash' has been explicitly marked deleted here
    __enum_hash() = delete;
    ^
```

**Comment by tedyu** [comment]
By accessing op->request().column_values(), we can get the column(s) for the insert operation.
e.g. inserting value 1 for column v:
```
ts2|pid1752|:16429|http://127.0.0.6:22424 I0814 18:02:31.066256 278376448 executor.cc:2118] type: 1 static ids:
ts2|pid1752|:16429|http://127.0.0.6:22424 I0814 18:02:31.066263 278376448 executor.cc:2122]  ids:
ts2|pid1752|:16429|http://127.0.0.6:22424 I0814 18:02:31.066267 278376448 executor.cc:2127]  column: column_id: 12 expr { value { int32_value: 1 } }
```
but the delete operation doesn't have this information.

**Comment by kmuthukk** [comment]
@tedyu  - My understanding was we want to disable batching for YCQL`begin transaction; stmt1; stmt2; end transaction;` statements. I am not clear why that change needs to deal with hashcode etc.?

**Comment by tedyu** [comment]
Currently Executor::WriteBatch::Add() uses the following to detect dependency between successive operations:
```
  if ((op->ReadsPrimaryRow() && ops_by_primary_key_.count(op) > 0) ||
```
ops_by_primary_key_ is an unordered_set.
Here are the relevant calls starting with count():
```
    size_type count(const key_type& __k) const {return __table_.__count_unique(__k);}
```
```
__hash_table<_Tp, _Hash, _Equal, _Alloc>::__count_unique(const _Key& __k) const
{
    return static_cast<size_type>(find(__k) != end());
```
```
__hash_table<_Tp, _Hash, _Equal, _Alloc>::find(const _Key& __k) const
{
    size_t __hash = hash_function()(__k);
```
Please see the last method - find where hash is involved.

From the test output, we know that ops_by_primary_key_.count(op) is 0 when adding the insert operation.
We may need to look at Executor::WriteBatch::Add() in more detail and see how the dependency detection can be made more robust.

**Comment by kmuthukk** [comment]
@tedyu -- I don't think we are trying to do any dependency detection. the plan is to keep this simple and unconditionally execute them one statement at a time, for now, since collisions can be not just on the primary row, but on the index too.

**Comment by tedyu** [comment]
Yeah - tuning dependency detection would take some time.

Let me try to modify Executor::WriteBatch::Add() where batching is disabled.

**Comment by tedyu** [comment]
I have a tentative fix which passes testInsertAfterDelete.
Still need to run through other tests in org.yb.cql.TestTransaction

**Comment by tedyu** [comment]
Original subject of the issue:

[YCQL] Run-time error on DELETE and INSERT on the same key in a transaction block

**Comment by tedyu** [comment]
The 'if not exists else error' clause corresponds to the following check in QLWriteOperation::Apply
```
    RETURN_NOT_OK(EvalCondition(request_.if_expr().condition(), existing_row, &should_apply));
    // Set the response accordingly.
    response_->set_applied(should_apply);
    if (!should_apply && request_.else_error()) {
      return ql::ErrorStatus(ql::ErrorCode::CONDITION_NOT_SATISFIED); // QLError
```

**Comment by OlegLoginov** [comment]
Fixed by  https://github.com/yugabyte/yugabyte-db/commit/6d66c9e6407ac5de9906096b9566635465bcb38f

**Comment by OlegLoginov** [comment]
Tested performance for the YCQL table INSERTs in a TRANSACTION:
```
SUMMARY:
--------
Experiment 1:  1 table + 1 index; TRANSACTION=4 INSERTs
Experiment 2:  2 tables + 2 indexes; TRANSACTION=2 INSERTs for tbl1 + 2 INSERTs for tbl2

FLAG: ycql_serial_operation_in_transaction_block:
        | Exp 1             | Exp 2
  FALSE | 2.24 ms/op        | 2.10 ms/op
  TRUE  | 3.88 ms/op (+73%) | 3.11 ms/op (+48%)
```

**Comment by OlegLoginov** [comment]
Details:
```
$ java -jar ./target/yb-sample-apps.jar --workload CassandraTransactionalKeyValue --nodes 127.0.0.1:9042 --num_threads_write 3 --num_threads_read 0 --num_unique_keys 1000000
ycql_serial_operation_in_transaction_block = TRUE

   80 [main] Connecting with 4 clients to nodes: /127.0.0.1:9042
  887 [main] Created a Cassandra table using query: [CREATE TABLE IF NOT EXISTS CassandraTransactionalKeyValue (k varchar, v bigint, primary key (k)) WITH transactions = { 'enabled' : true };]
 1578 [main] Created a Cassandra table using query: [CREATE INDEX ON CassandraTransactionalKeyValue(v);]
11590 [Thread-2] Read: 0.00 |  Write: 778.49 ops/sec  (3.85 ms/op), 9055 total ops 
16590 [Thread-2] Read: 0.00 |  Write: 905.86 ops/sec  (3.31 ms/op), 13585 total ops
21591 [Thread-2] Read: 0.00 |  Write: 876.29 ops/sec  (3.42 ms/op), 17967 total ops
26591 [Thread-2] Read: 0.00 |  Write: 772.94 ops/sec  (3.88 ms/op), 21832 total ops
31592 [Thread-2] Read: 0.00 |  Write: 715.75 ops/sec  (4.18 ms/op), 25411 total ops
36592 [Thread-2] Read: 0.00 |  Write: 943.92 ops/sec  (3.18 ms/op), 30131 total ops
41593 [Thread-2] Read: 0.00 |  Write: 735.54 ops/sec  (4.07 ms/op), 33809 total ops
46593 [Thread-2] Read: 0.00 |  Write: 854.52 ops/sec  (3.51 ms/op), 38082 total ops
51593 [Thread-2] Read: 0.00 |  Write: 652.75 ops/sec  (4.59 ms/op), 41346 total ops
3.18 - 4.59 ~ 3.88 ms/op

ycql_serial_operation_in_transaction_block = FALSE

  732 [main] Created a Cassandra table using query: [CREATE TABLE IF NOT EXISTS CassandraTransactionalKeyValue (k varchar, v bigint, primary key (k)) WITH transactions = { 'enabled' : true };]
 1126 [main] Created a Cassandra table using query: [CREATE INDEX IF NOT EXISTS ON CassandraTransactionalKeyValue(v);]
11155 [Thread-2] Read: 0.00 |  Write: 1818.83 ops/sec (1.65 ms/op), 17184 total ops
16156 [Thread-2] Read: 0.00 |  Write: 1707.49 ops/sec (1.75 ms/op), 25722 total ops
21156 [Thread-2] Read: 0.00 |  Write: 1601.10 ops/sec (1.87 ms/op), 33728 total ops
26157 [Thread-2] Read: 0.00 |  Write: 1754.02 ops/sec (1.71 ms/op), 42499 total ops
31157 [Thread-2] Read: 0.00 |  Write: 1638.47 ops/sec (1.83 ms/op), 50692 total ops
36157 [Thread-2] Read: 0.00 |  Write: 1825.32 ops/sec (1.64 ms/op), 59819 total ops
41158 [Thread-2] Read: 0.00 |  Write: 1602.89 ops/sec (1.87 ms/op), 67834 total ops
46158 [Thread-2] Read: 0.00 |  Write: 1700.28 ops/sec (1.76 ms/op), 76336 total ops
51159 [Thread-2] Read: 0.00 |  Write: 1030.48 ops/sec (2.84 ms/op), 81489 total ops
1.64 - 2.84 ~ 2.24 ms/op

=========================
ycql_serial_operation_in_transaction_block = TRUE

  816 [main] Created a Cassandra table using query: [CREATE TABLE IF NOT EXISTS CassandraTransactionalKeyValue (k varchar, v bigint, primary key (k)) WITH transactions = { 'enabled' : true };]
 1251 [main] Created a Cassandra table using query: [CREATE INDEX IF NOT EXISTS ON CassandraTransactionalKeyValue(v);]
 1529 [main] Created a Cassandra table using query: [CREATE TABLE IF NOT EXISTS CassandraTransactionalKeyValue2 (k varchar, v bigint, primary key (k)) WITH transactions = { 'enabled' : true };]
 1911 [main] Created a Cassandra table using query: [CREATE INDEX IF NOT EXISTS ON CassandraTransactionalKeyValue2(v);]
11944 [Thread-2] Read: 0.00 |  Write: 997.86 ops/sec  (3.00 ms/op), 9239 total ops  
16945 [Thread-2] Read: 0.00 |  Write: 867.09 ops/sec  (3.46 ms/op), 13575 total ops 
21945 [Thread-2] Read: 0.00 |  Write: 1058.36 ops/sec (2.83 ms/op), 18867 total ops
26945 [Thread-2] Read: 0.00 |  Write: 1024.95 ops/sec (2.92 ms/op), 23992 total ops
31946 [Thread-2] Read: 0.00 |  Write: 898.12 ops/sec  (3.34 ms/op), 28483 total ops 
36946 [Thread-2] Read: 0.00 |  Write: 858.71 ops/sec  (3.49 ms/op), 32777 total ops 
41947 [Thread-2] Read: 0.00 |  Write: 1021.13 ops/sec (2.94 ms/op), 37883 total ops
46947 [Thread-2] Read: 0.00 |  Write: 894.96 ops/sec  (3.35 ms/op), 42358 total ops 
51947 [Thread-2] Read: 0.00 |  Write: 917.14 ops/sec  (3.27 ms/op), 46944 total ops 
2.83 - 3.49 ~ 3.11

ycql_serial_operation_in_transaction_block = FALSE

  692 [main] Created a Cassandra table using query: [CREATE TABLE IF NOT EXISTS CassandraTransactionalKeyValue (k varchar, v bigint, primary key (k)) WITH transactions = { 'enabled' : true };]
 1058 [main] Created a Cassandra table using query: [CREATE INDEX IF NOT EXISTS ON CassandraTransactionalKeyValue(v);]
 1345 [main] Created a Cassandra table using query: [CREATE TABLE IF NOT EXISTS CassandraTransactionalKeyValue2 (k varchar, v bigint, primary key (k)) WITH transactions = { 'enabled' : true };]
 1761 [main] Created a Cassandra table using query: [CREATE INDEX IF NOT EXISTS ON CassandraTransactionalKeyValue2(v);]
11787 [Thread-2] Read: 0.00 |  Write: 1438.66 ops/sec (2.08 ms/op), 14136 total ops
16787 [Thread-2] Read: 0.00 |  Write: 1531.70 ops/sec (1.96 ms/op), 21795 total ops
21788 [Thread-2] Read: 0.00 |  Write: 1488.13 ops/sec (2.01 ms/op), 29236 total ops
26788 [Thread-2] Read: 0.00 |  Write: 1478.93 ops/sec (2.03 ms/op), 36631 total ops
31788 [Thread-2] Read: 0.00 |  Write: 1415.53 ops/sec (2.12 ms/op), 43709 total ops
36788 [Thread-2] Read: 0.00 |  Write: 1471.52 ops/sec (2.04 ms/op), 51067 total ops
41789 [Thread-2] Read: 0.00 |  Write: 1462.29 ops/sec (2.05 ms/op), 58379 total ops
46789 [Thread-2] Read: 0.00 |  Write: 1360.53 ops/sec (2.20 ms/op), 65182 total ops
51790 [Thread-2] Read: 0.00 |  Write: 1344.67 ops/sec (2.23 ms/op), 71906 total ops
1.96 - 2.23 ~ 2.10
```

**Comment by OlegLoginov** [comment]
As expected this feature increases transaction latency, but it has a small effect on overall CPU usage.
So assuming we/user increase number of threads/connections as needed they can get more/less similar throughput.
Alternatively, customers can just disable the serial processing using the flag `ycql_serial_operation_in_transaction_block = false`.
